File: elixir/examples/basics/main_test.exs
defmodule Convex.ExampleTest do
  use ExUnit.Case, async: true

  test "Float64 counter values produce the verifier's exact integer text" do
    assert Convex.Example.normalize_count!("initial", 0.0) == 0
    assert Convex.Example.normalize_count!("updated", 1.0) == 1

    assert "current count: #{Convex.Example.normalize_count!("query", 0.0)}" ==
             "current count: 0"

    assert "verified count: #{Convex.Example.normalize_count!("query", 0.0)} -> #{Convex.Example.normalize_count!("Live", 1.0)}" ==
             "verified count: 0 -> 1"
  end

  test "whole integer values remain valid" do
    assert Convex.Example.normalize_count!("query", 42) == 42
  end

  test "fractions and non-numeric values cannot look like a passing counter" do
    assert_raise RuntimeError, ~r/finite whole number/, fn ->
      Convex.Example.normalize_count!("query", 0.5)
    end

    assert_raise RuntimeError, ~r/finite whole number/, fn ->
      Convex.Example.normalize_count!("query", "0")
    end

    assert_raise RuntimeError, ~r/finite whole number/, fn ->
      Convex.Example.normalize_count!("query", :infinity)
    end
  end
end

File: java/examples/basics/MainTest.java
package convex.example;

import java.io.PrintWriter;
import java.io.StringWriter;

/** Checks the exact universal stdout transcript emitted by the canonical source. */
public final class MainTest {
  public static void main(String[] args) {
    StringWriter captured = new StringWriter();
    Main.writeTranscript(new PrintWriter(captured, true), 0, 0, true, 1, 1);
    String expected = "current count: 0\n" +
      "live initial count: 0\n" +
      "mutation applied: true\n" +
      "mutation count: 1\n" +
      "live updated count: 1\n" +
      "verified count: 0 -> 1\n";
    if (!expected.equals(captured.toString())) throw new AssertionError("unexpected example transcript:\n" + captured);
    System.out.println("example transcript test passed");
  }
}
